### DISCUSSION TITLE
Rollback batch optimistic update

### DISCUSSION THREAD

**Discussion by JesusTheHun** [TanStack/query#2570] [General] [answered]
I have several queries that have the same root : `['loadFoos', { filter1: 42 }]`, `['loadFoos', { filter2: 777 }]`. Unsurprisingly all those queries hold different data.

During the mutation, I would like to optimistically update all those data, which may or may not hold a reference to the mutated element. To update the data I simply call `queryClient.setQueryData()`, check if the mutated element is present, if so I update the query data.
The issue comes when I want to perform the rollback. 
Ideally I would get all the data from all the queries with `queryClient.getQueriesData()` before the optimistic update, and during the rollback I would reset the data to the collected one.
The problem is `queryClient.getQueriesData()` only returns the data, nothing about the query that holds it.
So when I want to rollback, I have no way to set the data back.

Right now I have found a workaround, it's more of a hack than anything else, but it works.

During the the optimistic update `queryClient.setQueriesData()` I store the new data in a map as the key, and the old data as the value. I return that map in the `context`. In the `onError` I do a `queryClient.setQueriesData()` again and search the reference of the new data in the map, get the value and set it as the old value. It works but it definitely feels like a hack.

Do you guys have come up with a better solution ?

**Comment by TkDodo** [comment, marked as answer]
> The problem is queryClient.getQueriesData() only returns the data, nothing about the query that holds it.

Hm, I don't think that's true. `getQueriesData` returns: `[QueryKey, TData][]`

so An Array of tuples, each having the exact query key and their data, for this exact reason.

**Reply by JesusTheHun** [reply]
Looks like I should have taken a walk ... thanks !
